system-helper/flatpak-system-helper.c

```diff
@@ -190,6 +420,341 @@ get_sender_pid (GDBusMethodInvocation *invocation)
+        {
+          g_autoptr(GFile) cache_dir = g_file_get_child (repo_tmpfile, name);
+
+          if (check_for_repo_containing_partial_pull (cache_dir, commit, out_location))
+            return TRUE;
+        }
+    }
+
+  return FALSE;
+}
+
+static gboolean
+handle_get_child_repo_for_pull (FlatpakSystemHelper   *object,
+                                GDBusMethodInvocation *invocation,
+                                const gchar           *ref,
+                                const gchar           *commit,
+                                const gchar           *arg_installation)
+{
+  GDBusConnection *connection = g_dbus_method_invocation_get_connection (invocation);
+  g_autoptr(FlatpakDir) system = NULL;
+  g_autoptr(GError) error = NULL;
+  g_autofree gchar *bindfs_path = NULL;
+
+  g_debug ("GetChildRepoForPull %s", arg_installation);
+
+  system = dir_get_system (arg_installation, &error);
+  if (system == NULL)
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return TRUE;
+    }
+
+  bindfs_path = g_find_program_in_path ("bindfs");
+  if (bindfs_path == NULL)
+    {
+      g_dbus_method_invocation_return_error (invocation, G_DBUS_ERROR, G_DBUS_ERROR_NOT_SUPPORTED,
+                                             "Can't find bindfs in the system.");
+      return TRUE;
+    }
+  else
+    {
+      g_autoptr(FlatpakSystemHelperOngoingPull) new_pull = NULL;
+      g_autofree gchar *child_repo_path = NULL;
+      g_autofree gchar *user_dir = NULL;
+      g_autofree gchar *root_dir = NULL;
+      g_autofree gchar *flatpak_dir = NULL;
+      g_autofree gchar *arg_user = NULL;
+      g_autofree gchar *out_location = NULL;
+      guint uid_pull_count = 0;
+      uid_t uid;
+
+      if (!get_connection_uid (invocation, &uid, &error))
+        {
+          g_dbus_method_invocation_return_gerror (invocation, error);
+          return TRUE;
+        }
+
+      uid_pull_count = get_uid_pull_count (uid);
+      if (uid_pull_count > MAX_PULLS_PER_USER)
+        {
+          g_dbus_method_invocation_return_error (invocation, G_DBUS_ERROR, G_DBUS_ERROR_LIMITS_EXCEEDED,
+                                                 "Maximum number of pulls per user exceeded for UID: %u", uid);
+          return TRUE;
+        }
+
+      flatpak_dir = g_file_get_path (flatpak_dir_get_path (system));
+      g_autofree gchar *repo_tmp = g_build_filename (flatpak_dir, "repo", "tmp", NULL);
+
+      g_auto(GStrv) parts = flatpak_decompose_ref (ref, NULL);
+      if (check_partial_pull_availability (repo_tmp, parts[1], commit, &out_location))
+        {
+          g_autoptr(GFile) user_dir_file = NULL;
+
+          g_debug ("Found a partial pull for ref %s at: %s", parts[1], out_location);
+          root_dir = g_path_get_dirname (out_location);
+          user_dir = g_strdup_printf ("%s-bindfs", root_dir);
+          user_dir_file = g_file_new_for_path (user_dir);
+          if (!g_file_query_exists (user_dir_file, NULL))
```

Comment:
racy. maybe it doesn't matter. but what if the directory already exists. thats ok ?

Reply:
Yeah, I read it. I think it doesn't matter here.
Yes, if the directory already exists, it's OK. We will reuse it to mount it. Otherwise this directory(i.e. `user_dir`) is missing we will recreate it (although this is unlikely).